Add Beatmap Information elements

C0D3-M4513R opened the pull request:
This pr focuses on this idea: https://github.com/ppy/osu/discussions/20608#discussioncomment-4028299

Essentially I want to add the current Beatmap information to the Osu SkinEditor, so that you could make a Streaming setup without having to have a seperate Programm open (like you have to in stable).
Because of that intent I also added a TextElement, which basically allows you to add arbitrary Text on the screen. (why wasn't this included earlier?)

I can imagine that making the `Beatmap` property of `OsuBaseGame` public is probably not something looked favorably upon, but please leave suggestion on how else I can do it.

C0D3-M4513R committed:
Initial implementation of a Beatmap Information Skinning Element

Signed-off-by: C0D3 M4513R <28912031+C0D3-M4513R@users.noreply.github.com>

C0D3-M4513R committed:
Add a generic text Element

Signed-off-by: C0D3 M4513R <28912031+C0D3-M4513R@users.noreply.github.com>

bdach commented:
I'm pretty sure that the intended direction for this sort of thing was to have the game expose a read-only API that could be consumed by all sorts of third-party apps rather than jury-rig the skinning system to do this.

See also:

- https://github.com/ppy/osu/discussions/13011
- https://github.com/ppy/osu/discussions/16949
- https://github.com/ppy/osu/discussions/15061
- https://github.com/ppy/osu/pull/18129

So unfortunately, I'm sorry but I don't think this PR is going to get in.

bdach commented:
The issue isn't "the way (you) read the current beatmap". The issue is that to the best to my knowledge we do not want to add the types of skinning elements that you add in this pull *at all*. If the websocket API linked above happens, then consumers can write their own external tools that interact with it for their OBS layout or whatever, *as long as that is done outside of the game itself*.

C0D3-M4513R committed:
Don't make Beatmap in OsuGameBase public

Signed-off-by: C0D3 M4513R <28912031+C0D3-M4513R@users.noreply.github.com>

smoogipoo commented:
I agree that this PR and the websocket PR are completely different ideas/uses. Am not closing it for the time being.

peppy commented on osu.Game/Skinning/Components/BeatmapInfoDrawable.cs:
```diff
@@ -0,0 +1,180 @@
+                case BeatmapInfo.Length:
+                    beatmap.BindValueChanged(bm =>
+                    {
+                        const long ms_to_s = 1000;
```
`ToFormattedDuration` might work?

https://github.com/ppy/osu/blob/f8830c6850128456266c82de83273204f8b74ac0/osu.Game/Overlays/BeatmapSet/BasicStats.cs#L71-L72

peppy commented on osu.Game/Skinning/Components/BeatmapInfoDrawable.cs:
```diff
@@ -0,0 +1,180 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
Please remove these and adjust classes. We don't add NRT disable to new classes.

peppy commented:
I like this direction, but I think that labels should be included for these displays, rather than leaving that up to the user to implement. This would also allow for it to be localised.

If you want to make the label customisable (please provide reasoning if so) then there could be an override text setting on the `BeatmapInfoDrawable` rather than using a completely separate component. It could be empty by default and when empty, would use the default label.

peppy commented:
> The issue isn't "the way (you) read the current beatmap". The issue is that to the best to my knowledge we do not want to add the types of skinning elements that you add in this pull _at all_. If the websocket API linked above happens, then consumers can write their own external tools that interact with it for their OBS layout or whatever, _as long as that is done outside of the game itself_.

I think it's fine to add these kinds of components. They are easy to add, maintain, and allow users a lot of flexibility to how they layout their screens. Eventually these could replace the song select fixed implementations to allow rearranging data display there, too.

C0D3-M4513R committed:
Add Localisation

Also add Labels for what is displayed, and prefix/suffix for Labels
Add a Prefix and Suffix for Values

C0D3-M4513R committed:
Remove nullable disable from TextElement

C0D3-M4513R committed:
Use more Localisations

C0D3-M4513R committed:
Fix Skin Deserialisation Test

C0D3-M4513R committed:
Merge branch 'master' into beatmap-info

peppy commented on osu.Game/Skinning/Components/BeatmapInfoDrawable.cs:
```diff
@@ -0,0 +1,314 @@
+            switch (type.NewValue)
+            {
+                case BeatmapInfo.CircleSize:
+                    beatmap.BindValueChanged(bm =>
```
I believe you're leaking events here, since every time the `Type` is changed you're adding a new `ValueChanged`. A single line fix would be to call `beatmap.UnbindAllEvents()` at the start of this method. A more correct fix would probably be to switch the nesting here so the `switch` is done inside a single `beatmap.BindValueChanged()` which is bound in `LoadComplete`. Something like this:

```csharp
LoadComplete()
{
    type.BindValueChanged(_ => updateBeatmapContent());
    beatmap.BindValueChanged(_ => updateBeatmapContent(), true);
}

private updateBeatmapContent()
{
    switch (type.Value)
    {
        ...
    }
}
```

C0D3-M4513R committed:
Fix potential resource leak

C0D3-M4513R committed:
Merge branch 'master' into beatmap-info

C0D3-M4513R committed:
Allow for the Value of BeatmapInfoDrawable to be formatted

C0D3-M4513R committed:
Merge remote-tracking branch 'origin/beatmap-info' into beatmap-info

bdach commented:
The scope creep here severely worries me and is part of why I didn't think this belongs in the game in the first place. Now there's customisable labels, templating logic, custom suffixes, prefixes...

I'm personally looking away from this pull. It feels like a bad path to go down.

C0D3-M4513R committed:
Merge branch 'master' into beatmap-info

peppy commented:
I wrote up a response to this at some point but never posted it. I tend to agree that the user should not have to type in template names to make this work, but as long as the dropdown prefills the template field then it may work okay.

I think if you're going in this direction, there should only be a single text field. the whole prefix/suffix/whatever logic should be removed and everything should be moved to the "Value Formatting" field (and that should be renamed to "Template" or something).

peppy commented:
Another way of thinking about it, to give better structure would be to have the dropdown decide *all inputs* to the format string, then use standard c# index based formatting.

ie. select "Artist / Title" from dropdown, then the format string is prefilled with `{0} - {1}` which would be `Aritst - Title`.

This would obviously give the user less flexibility, but may appease the issue of scope creep in this feature.

peppy reviewed (commented):
As proposed.

C0D3-M4513R committed:
Simplify the whole Templating process

C0D3-M4513R committed:
Use AutoSizeAxes

C0D3-M4513R committed:
Merge branch 'master' into beatmap-info

C0D3-M4513R committed:
Merge branch 'master' into beatmap-info

peppy committed:
Merge branch 'master' into beatmap-info

peppy committed:
Rename class and reword settings text/tooltips to avoid clashing with other naming

peppy committed:
Rename attributes in enum to match better with user expectations

peppy committed:
Move label initialisation out of ctor

peppy committed:
Use `LocalisableDescription` from `BeatmapOnlineStatus` enum instead of locally defined

peppy committed:
Simplify attribute assignment

peppy committed:
Simplify class structure

peppy committed:
Update serialisation test to match new comopnent naming

peppy commented:
@C0D3-M4513R I've done quite a bit of clean-up work on this PR, and think, it's in a good state now. Please check through my commits to see what was required to get it up-to-scratch, and ask if anything stands out.

Would appreciate a final review on this before merging.

peppy commented:
Would appreciate one more review on this since I made considerable changes.

bdach commented on osu.Game/Skinning/Components/BeatmapAttributeText.cs:
```diff
@@ -0,0 +1,133 @@
+
+        private void updateLabel()
+        {
+            string newText = Template.Value.Replace("{Label}", label_dictionary[Attribute.Value].ToString())
+                                     .Replace("{Value}", valueDictionary[Attribute.Value].ToString());
+
+            foreach (var type in Enum.GetValues(typeof(BeatmapAttribute)).Cast<BeatmapAttribute>())
+            {
+                newText = newText.Replace("{" + type + "}", valueDictionary[type].ToString());
+            }
+
+            text.Text = newText;
```
The rest of this class gives off the appearance that this should support localisation, but it actually doesn't, due to the `.ToString()` calls here that just end up using fallback English values. Was the intent here to support localisation fully?

If yes, a *rough* implementation of this method that works with localisation follows (although I don't claim it to be a particularly smart or good one) follows:

```diff
diff --git a/osu.Game/Skinning/Components/BeatmapAttributeText.cs b/osu.Game/Skinning/Components/BeatmapAttributeText.cs
index 572b4ba818..69136b3bb1 100644
--- a/osu.Game/Skinning/Components/BeatmapAttributeText.cs
+++ b/osu.Game/Skinning/Components/BeatmapAttributeText.cs
@@ -103,15 +103,28 @@ private void updateBeatmapContent(WorkingBeatmap workingBeatmap)
 
         private void updateLabel()
         {
-            string newText = Template.Value.Replace("{Label}", label_dictionary[Attribute.Value].ToString())
-                                     .Replace("{Value}", valueDictionary[Attribute.Value].ToString());
+            string numberedTemplate = Template.Value.Replace("{Label}", "{0}")
+                                              .Replace("{Value}", $"{{{1 + (int)Attribute.Value}}}");
 
             foreach (var type in Enum.GetValues(typeof(BeatmapAttribute)).Cast<BeatmapAttribute>())
             {
-                newText = newText.Replace("{" + type + "}", valueDictionary[type].ToString());
+                numberedTemplate = numberedTemplate.Replace("{" + type + "}", $"{{{1 + (int)type}}}");
             }
 
-            text.Text = newText;
+            var args = valueDictionary.OrderBy(pair => pair.Key)
+                                      .Select(pair => pair.Value)
+                                      .Prepend(label_dictionary[Attribute.Value])
+                                      .Cast<object?>()
+                                      .ToArray();
+
+            try
+            {
+                text.Text = LocalisableString.Format(numberedTemplate, args);
+            }
+            catch
+            {
+                text.Text = "Invalid format!";
+            }
         }
     }

 ```

The try-catch at the end is an escape hatch - because with this implementation user input is more or less directly translated to `string.Format` calls, trying to input an open brace would crash because `{` is not a valid format string spec for `string.Format` (it needs to be either terminated as a placeholder, or if a literal brace was meant, it needs to be doubled up as `{{` to be escaped).

bdach commented on osu.Game/Skinning/Components/BeatmapAttributeText.cs:
```diff
@@ -0,0 +1,133 @@
+            valueDictionary[BeatmapAttribute.Creator] = workingBeatmap.BeatmapInfo.Metadata.Author.Username;
+            valueDictionary[BeatmapAttribute.Length] = TimeSpan.FromMilliseconds(workingBeatmap.BeatmapInfo.Length).ToFormattedDuration();
+            valueDictionary[BeatmapAttribute.RankedStatus] = workingBeatmap.BeatmapInfo.Status.GetLocalisableDescription();
+            valueDictionary[BeatmapAttribute.BPM] = workingBeatmap.BeatmapInfo.BPM.ToString(@"F2");
+            valueDictionary[BeatmapAttribute.CircleSize] = ((double)workingBeatmap.BeatmapInfo.Difficulty.CircleSize).ToString(@"F2");
+            valueDictionary[BeatmapAttribute.HPDrain] = ((double)workingBeatmap.BeatmapInfo.Difficulty.DrainRate).ToString(@"F2");
+            valueDictionary[BeatmapAttribute.Accuracy] = ((double)workingBeatmap.BeatmapInfo.Difficulty.OverallDifficulty).ToString(@"F2");
+            valueDictionary[BeatmapAttribute.ApproachRate] = ((double)workingBeatmap.BeatmapInfo.Difficulty.ApproachRate).ToString(@"F2");
+            valueDictionary[BeatmapAttribute.StarRating] = workingBeatmap.BeatmapInfo.StarRating.ToString(@"F2");
```
Similarly, was the intent here to call `.ToString()` rather than `.ToLocalisableString()`?

C0D3-M4513R committed:
Use LocaliseableStings in `BeatmapAttributeText`

peppy commented:
Seems to work as expected. Let's get this in and save any further work for follow-up PRs.
